src/engine/GroupBy.cpp

```diff
@@ -677,6 +697,123 @@ bool GroupBy::computeOptimizedGroupByIfPossible(IdTable* result) {
+  // TODO: Use an interface for AggregationData
+  ad_utility::HashMap<ValueId, AverageAggregationData> map;
+
+  // Initialize evaluation context
+  sparqlExpression::EvaluationContext evaluationContext(
+      *getExecutionContext(), _subtree->getVariableColumns(),
+      subresult,getExecutionContext()->getAllocator(), localVocab);
+
+  evaluationContext._groupedVariables = ad_utility::HashSet<Variable>{
+      _groupByVariables.begin(), _groupByVariables.end()};
+  // TODO: Save previous results (?)
+  evaluationContext._variableToColumnMapPreviousResults =
+      getInternallyVisibleVariableColumns();
+  evaluationContext._previousResultsFromSameGroup.resize(getResultWidth());
+  evaluationContext._isPartOfGroupBy = true;
+
+  for (size_t i = 0; i < subresult.size(); ++i) {
+    auto id = subresult(i, columnIndex);
+
+    // We always calculate for a single row
+    evaluationContext._beginIndex = i;
+    evaluationContext._endIndex = i + 1;
+
+    ValueId resultEntry;
+
+    // Evaluate child expression
+    auto expr = aggregates.at(0)._expression.getPimpl();
+    auto exprChildren = expr->children();
+
+    sparqlExpression::ExpressionResult expressionResult =
+        exprChildren[0]->evaluate(&evaluationContext);
+
+    // Extract ValueId
+    auto visitor = [&]<sparqlExpression::SingleExpressionResult T>(
+                       T&& singleResult) mutable {
+      // If the expression is a variable, get value directly
+      constexpr static bool isVariable = std::is_same_v<T, Variable>;
+      if constexpr (isVariable) {
+        std::span<const ValueId> results = sparqlExpression::detail::getIdsFromVariable(singleResult,
+                                                                                        &evaluationContext);
+        resultEntry = results.front();
+        return;
+      }
+      constexpr static bool isStrongId = std::is_same_v<T, Id>;
+      AD_CONTRACT_CHECK(sparqlExpression::isConstantResult<T> ||
+                        sparqlExpression::isVectorResult<T>);
+      if constexpr (isStrongId) {
+        resultEntry = singleResult;
+      } else if constexpr (sparqlExpression::isConstantResult<T>) {
+        resultEntry = sparqlExpression::detail::constantExpressionResultToId(
+            AD_FWD(singleResult), localVocab);
+      } else if constexpr (sparqlExpression::isVectorResult<T>) {
+        resultEntry = sparqlExpression::detail::constantExpressionResultToId(
+            std::move(singleResult[0]), localVocab);
+      } else {
+        // Can this happen? What expressions return something other than
+        // constants or vectors with a single element?
+        AD_FAIL();
+      }
+    };
+
+    std::visit(visitor, std::move(expressionResult));
+
+    // TODO: Check datatype, use getXYZ and cast accordingly
+    if (map.contains(id)) {
+      map.at(id).increment(resultEntry.getInt());
+    } else {
+      map.insert({id, {(double) resultEntry.getInt(), 1}});
+    }
+  }
```

Comment:
Small trick:
If the default constructor of the types in the map do the right thing (e.g. initialize to zero), then you can simply write `map[id].increment(....)` instead of the `if-else`.